src/main/host/syscall/handler/mod.rs

```diff
@@ -581,23 +581,23 @@ impl SyscallHandler {
                 // only show a warning the first time we encounter this unsupported syscall
                 static WARNED_SET: RwLock<Option<HashSet<SyscallNum>>> = RwLock::new(None);
 
+                // First check with a (cheap) read-lock.
                 let has_already_warned = WARNED_SET
                     .read()
                     .unwrap()
                     .as_ref()
                     .map(|x| x.contains(&syscall))
                     .unwrap_or(false);
 
-                if !has_already_warned {
-                    // `insert()` returns `false` if the syscall num was already in the set.
-                    // This can happen if another thread added the syscall after we released
-                    // the read-lock, above, and took the write lock, below.
-                    WARNED_SET
+                // If it looks like we haven't already warned, add this to the
+                // already-warned set. Also detect the (rare) case that another
+                // thread already warned after we released the read-lock above.
+                let has_already_warned = has_already_warned
+                    || WARNED_SET
```

Comment:
This seems backwards to me? `insert` should return true when it's added for the first time, which would mean `has_already_warned` should be false.

Reply:
Doh, right. Will follow up with another fix.

Whenever I look at this  code I'm tempted to just use a Mutex instead of a RwLock since it *shouldn't* be on the hot path, but always end up deciding "oh well, this one last tweak should fix it once and for all anyway"

Reply:
Yeah it originally used a mutex, but then was changed to RwLock during review. As far as I remember there wasn't a strong argument for either one.

I did refactor this into a separate module as part of https://github.com/shadow/shadow/pull/3332, but I'm not sure if/when I'll merge that.